### PR TITLE
Implement `integer, parameter :: x(*) = [1, 2, 3]`

### LINKED ISSUE #967: Implement `integer, parameter :: x(*) = [1, 2, 3]`
Example:
```fortran
program const_arrays
integer, parameter :: x(*) = [1, 2, 3]
print *, x
end
```
Currently one gets:
```console
$ lfortran arrays_01.f90 
Traceback (most recent call last):
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 1794
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 774
    res = fe.get_llvm3(*asr, lpm, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/fortran_evaluator.cpp", line 321
    compiler_options, run_fn);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6195
    v.nested_func_types = pass_find_nested_vars(asr, context,
  File "../libasr/asr.h", line 4080
  File "../libasr/asr.h", line 4057
  File "../libasr/asr.h", line 4081
  File "../libasr/asr.h", line 3836
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 1157
    visit_symbol(*item.second);
  File "../libasr/asr.h", line 4083
  File "../libasr/asr.h", line 3845
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 2160
    declare_vars(x);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 2505
    fill_array_details(ptr, llvm_data_type, m_dims, n_dims);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 426
    visit_expr(*(m_dim.m_start));
  File "../libasr/asr.h", line 4141
  File "../libasr/asr.h", line 4141
  Binary file "/usr/lib/system/libsystem_platform.dylib", local address: 0x1803244e3
Segfault: Signal SIGSEGV (segmentation fault) received
```

I don't think it's difficult to fix, we just need to get the compile time length of the RHS, and use that as the length of the LHS.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Resolves #967

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -803,7 +804,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
     }
 
     void process_dims(Allocator &al, Vec<ASR::dimension_t> &dims,
-        AST::dimension_t *m_dim, size_t n_dim) {
+        AST::dimension_t *m_dim, size_t n_dim, int &is_compile_time=temp_value) {
```
Comment (czgdp1807) [inline src/lfortran/semantics/ast_common_visitor.h:807]: Why can't we make `is_compile_time` a compulsory argument? I guess we will just need to update calls to `process_dims` calls which is fine.

File: tests/tests.toml [hunk 2]
```diff
@@ -2594,4 +2594,9 @@ asr = true
...
 [[test]]
 filename = "../integration_tests/implied_do_loops3.f90"
-asr = true
\ No newline at end of file
+asr = true
+
+[[test]]
+filename = "array10.f90"
+asr = true
+llvm = true
```
Comment (czgdp1807) [inline tests/tests.toml:2602]:
```suggestion
```

LLVM not needed. Already running via `integration_test/const_array_01.f90`

File: tests/array10.f90 [hunk 3]
```diff
@@ -0,0 +1,4 @@
+program const_arrays
+	integer, parameter :: x(*) = [4,9,6,12,13]
+	print *, x
+end
```
Comment (czgdp1807) [inline tests/array10.f90:4]:
```suggestion
program const_arrays
	integer, parameter :: x(*) = [4, 9, 6, 12, 13]
	print *, x
end

```

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -803,7 +803,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
     }
 
     void process_dims(Allocator &al, Vec<ASR::dimension_t> &dims,
-        AST::dimension_t *m_dim, size_t n_dim) {
+        AST::dimension_t *m_dim, size_t n_dim, int &is_compile_time) {
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:806]:
```suggestion
        AST::dimension_t *m_dim, size_t n_dim, bool &is_compile_time) {
```
Also it is not initialized in all cases. You have to set `is_compile_time = false` at the beginning of the function.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 5]
```diff
@@ -1145,10 +1148,11 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                                 throw SemanticError("Cannot set dimension for undeclared variable", x.base.base.loc);
                             }
                         }
+                        int is_compile_time = 0;
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1151]:
```suggestion
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 6]
```diff
@@ -1145,10 +1148,11 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
                         if (ASR::is_a<ASR::Variable_t>(*get_sym)) {
                             Vec<ASR::dimension_t> dims;
                             dims.reserve(al, 0);
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1154]: Add
```c++
                            bool is_compile_time;
```
here.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 7]
```diff
@@ -1172,6 +1176,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
             // Example
             // real(dp), private :: x, y(3), z
             for (size_t i=0; i<x.n_syms; i++) {
+                int is_compile_time = 0;
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1179]:
```suggestion
                bool is_compile_time = false;
```

**Review by certik — CHANGES_REQUESTED** [review]
Otherwise looks good I think. Thank for fixing it!

**Review by czgdp1807 — APPROVED** [review]

**Review by certik — APPROVED** [review]
Thanks!

**Pull request merged**